File: pathfinder-frontend/src/v7/ConditionalComponent.ts
import {Choice} from "./Choice";
import {Effect} from "./Effect";
import {FeatureResolvable} from "./FeatureResolvable";
import Link from "./Link";
import {ResolvedEntityContext} from "./ResolvedEntityContext";

export default class ConditionalComponent implements FeatureResolvable {

  constructor(public readonly conditionFormula: string,
              public readonly effects: Effect[],
              public readonly links: Link[],
              public readonly choices: Choice[]) {
  }

  async resolve(basePath: string, context: ResolvedEntityContext): Promise<void> {
    for (const link of this.links) {
      await link.resolve(basePath, context);
    }
    for (const choice of this.choices) {
      await choice.resolve(basePath, context);
    }
  }
}

File: pathfinder-generator/src/main/java/pathfinder/model/ConditionalComponent.java
package pathfinder.model;

import static pathfinder.util.ListUtils.mapList;

import pathfinder.data.ConditionalStackComponentDbo;

public record ConditionalComponent(String conditionFormula,
                                   Component component) {

    public ConditionalStackComponentDbo toDbo() {
        return ConditionalStackComponentDbo.newBuilder()
                .setConditionFormula(conditionFormula)
                .addAllEffects(mapList(component.effects(), Effect::toDbo))
                .addAllLinks(mapList(component.links(), Link::toDbo))
                .addAllUnlinks(mapList(component.unlinks(), Unlink::toDbo))
                .addAllChoices(mapList(component.choices(), Choice::toDbo))
                .build();
    }
}
